src/webgpu/shader/execution/memory_model/texture_intra_invocation_coherence.spec.ts

```diff
@@ -0,0 +1,331 @@
+    }
+  }
+  t += `<${format}, read_write>`;
+  return t;
+}
+
+function textureStore(dim: GPUTextureViewDimension, index: string): string {
+  let code = `textureStore(t, indexToCoord(${index}), `;
+  if (dim === '2d-array') {
+    code += `0, `;
+  }
+  code += `texel)`;
+  return code;
+}
+
+function textureLoad(dim: GPUTextureViewDimension, format: GPUTextureFormat): string {
+  let code = ``;
+  if (format !== 'r32uint') {
+    code += `u32(`;
+  }
+  code += `textureLoad(t, indexToCoord(read_index[global_index])`;
+  if (dim === '2d-array') {
+    code += `, 0`;
+  }
+  code += `).x`;
+  if (format !== 'r32uint') {
+    code += `)`;
+  }
+  return code;
+}
+
+function texel(format: GPUTextureFormat): string {
+  switch (format) {
+    case 'r32uint': {
+      return 'vec4u(global_index,0,0,0)';
+    }
+    case 'r32sint': {
+      return 'vec4i(i32(global_index),0,0,0)';
+    }
+    case 'r32float': {
+      return 'vec4f(f32(global_index),0,0,0)';
+    }
+    default: {
+      unreachable('unhandled format: ${format}');
+    }
+  }
+  return '';
+}
+
+function getTextureSize(numTexels: number, dim: GPUTextureViewDimension): GPUExtent3D {
+  const size: GPUExtent3D = { width: 1, height: 1, depthOrArrayLayers: 1 };
+  switch (dim) {
+    case '1d': {
+      size.width = numTexels;
+      break;
+    }
+    case '2d':
+    case '2d-array':
+    case '3d': {
+      size.width = numTexels / 2;
+      size.height = numTexels / 2;
+      // depthOrArrayLayers defaults to 1
+      break;
+    }
+    default: {
+      unreachable(`unhandled dim: ${dim}`);
+    }
+  }
+  return size;
+}
+
+g.test('texture_intra_invocation_coherence')
+  .desc(`Tests writes from an invocation are visible to reads from the same invocation`)
+  .params(u => u.combine('format', kRWStorageFormats).combine('dim', kDimensions))
+  .beforeAllSubcases(t => {
+    t.selectDeviceForTextureFormatOrSkipTestCase(t.params.format);
+  })
+  .fn(t => {
+    t.skipIfLanguageFeatureNotSupported('readonly_and_readwrite_storage_textures');
+
+    const wgx = 16;
+    const wgy = 16;
+    const num_wgs_x = 2;
+    const num_wgs_y = 2;
+    const invocations = wgx * wgy * num_wgs_x * num_wgs_y;
+
+    const code = `
+requires readonly_and_readwrite_storage_textures;
+
+@group(0) @binding(0)
+var t : ${textureType(t.params.format, t.params.dim)};
+
+@group(1) @binding(0)
+var<storage> write_indices : array<vec4u>;
+
+@group(1) @binding(1)
+var<storage> read_index : array<u32>;
+
+@group(1) @binding(2)
+var<storage> write_mask : array<vec4u>;
+
+@group(1) @binding(3)
+var<storage, read_write> output : array<u32>;
+
+const wgx = ${wgx}u;
+const wgy = ${wgy}u;
+const num_wgs_x = ${num_wgs_x}u;
+const num_wgs_y = ${num_wgs_y}u;
+
+${indexToCoord(t.params.dim)}
+
+@compute @workgroup_size(wgx, wgy, 1)
+fn main(@builtin(local_invocation_index) lid : u32,
+        @builtin(workgroup_id) wgid : vec3u) {
+  let global_index = wgid.y * num_wgs_x * wgx * wgy + wgid.x * wgx * wgy + lid;
```

Comment:
nit: this could be simpler, by using the global_invocation_id  builtin.

`gid.x + gid.y * wgx + gid.z * wgx * wgy`